# Setup
Initial Setup and Versioning

In [40]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [41]:
import os
from datetime import datetime
import json

In [42]:
# Version info
MODEL_VERSION = "Llama-3.2-1B"
TIMESTAMP = datetime.now().strftime("%Y%m%d_%H%M")

# Create versioned output directories
base_output_dir = "/content/drive/MyDrive/outputs/llama_fine_tuning"
model_output_dir = f"{base_output_dir}/{MODEL_VERSION}/model_{TIMESTAMP}"
profiles_output_dir = f"{base_output_dir}/{MODEL_VERSION}/profiles_{TIMESTAMP}"
logs_dir = f"{base_output_dir}/{MODEL_VERSION}/logs_{TIMESTAMP}"

# Create directories if they don't exist
for dir_path in [model_output_dir, profiles_output_dir, logs_dir]:
    os.makedirs(dir_path, exist_ok=True)

# Save version info
version_info = {
    "model_version": MODEL_VERSION,
    "timestamp": TIMESTAMP,
    "base_model": "Llama-3.2-1B",
    "training_data": "original_dataset",
    "description": "Initial fine-tuning run with original dataset"
}

with open(f"{logs_dir}/version_info.json", 'w') as f:
    json.dump(version_info, f, indent=2)

print(f"Initialized version {MODEL_VERSION} at {TIMESTAMP}")
print(f"Model outputs will be saved to: {model_output_dir}")
print(f"Generated profiles will be saved to: {profiles_output_dir}")

Initialized version Llama-3.2-1B at 20241118_1824
Model outputs will be saved to: /content/drive/MyDrive/outputs/llama_fine_tuning/Llama-3.2-1B/model_20241118_1824
Generated profiles will be saved to: /content/drive/MyDrive/outputs/llama_fine_tuning/Llama-3.2-1B/profiles_20241118_1824


# Data Loading and Preprocessing

- 8-bit quantization
- LoRA for efficient fine-tuning
- Gradient checkpointing

In [43]:
!pip install datasets

In [44]:
import logging
from datasets import Dataset
from transformers import AutoTokenizer

In [45]:
# Setup logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler(f"{logs_dir}/training.log"),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)

logger.info("Starting data loading and preprocessing")

def load_dataset():
    # Load the JSON data
    with open("/content/drive/MyDrive/data/processed/llama_training/processed_original_dataset.json", 'r') as f:
        data = json.load(f)
    logger.info(f"Loaded {len(data)} examples from dataset")
    return data

def prepare_dataset(data):
    # Create a HuggingFace dataset
    dataset = Dataset.from_dict({
        "text": [item["text"] for item in data]
    })
    logger.info("Created HuggingFace dataset")
    return dataset

# Load and prepare the data
raw_data = load_dataset()
dataset = prepare_dataset(raw_data)

# Split into train and validation sets
train_val_dataset = dataset.train_test_split(test_size=0.1, seed=42)
logger.info(f"Training set size: {len(train_val_dataset['train'])}")
logger.info(f"Validation set size: {len(train_val_dataset['test'])}")

# Initialize tokenizer
logger.info("Initializing tokenizer...")
tokenizer = AutoTokenizer.from_pretrained(
    "meta-llama/Llama-3.2-1B",
    token="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",
    trust_remote_code=True
)
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right"

# Print a sample to verify data format
print("\nSample from dataset:")
print(dataset[0]['text'][:500] + "...")  # Print first 500 chars of first example


Sample from dataset:
### Instruction:
Generate a detailed sperm donor profile based on these characteristics:
Height: 5'10 (178cm)
Weight: 162 lbs (73kg)
Eye Color: Black
Hair: Dark Brown
Education: Master in Architecture
Ethnic Background: East Indian

### Response:
Quadrilingual Architect. loves his career as an architect. He’s had a lifelong creative streak and drawing has always been a favorite pastime, from doodling to designing buildings! This smart cookie has an M.S. in architecture (3.7 GPA) and can even spe...


### Cell Output

Sample from dataset:
### Instruction:
Generate a detailed sperm donor profile based on these characteristics:
Height: 5'10 (178cm)
Weight: 162 lbs (73kg)
Eye Color: Black
Hair: Dark Brown
Education: Master in Architecture
Ethnic Background: East Indian

### Response:
Quadrilingual Architect. loves his career as an architect. He’s had a lifelong creative streak and drawing has always been a favorite pastime, from doodling to designing buildings! This smart cookie has an M.S. in architecture (3.7 GPA) and can even spe...

# Model Initiation

In [46]:
!pip install -U bitsandbytes

In [47]:
!pip show bitsandbytes


Name: bitsandbytes
Version: 0.44.1
Summary: k-bit optimizers and matrix multiplication routines.
Home-page: https://github.com/TimDettmers/bitsandbytes
Author: Tim Dettmers
Author-email: dettmers@cs.washington.edu
License: MIT
Location: /usr/local/lib/python3.10/dist-packages
Requires: numpy, torch
Required-by: 


In [48]:
import torch
from transformers import AutoModelForCausalLM, BitsAndBytesConfig
from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training

In [49]:
import logging

# Set up the logger
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


In [50]:
logger.info("Starting model initialization...")

In [51]:
# Memory optimization settings
logger.info("Setting up memory optimizations...")
torch.backends.cuda.matmul.allow_tf32 = True
torch.backends.cudnn.allow_tf32 = True

# Configure quantization
quantization_config = BitsAndBytesConfig(
    load_in_8bit=True,
    llm_int8_threshold=6.0,
    llm_int8_has_fp16_weight=True
)

In [52]:
# Print available GPU memory before model loading
if torch.cuda.is_available():
    logger.info(f"Available GPU memory before model loading: {torch.cuda.get_device_properties(0).total_memory / 1e9:.2f} GB")


In [53]:
tokenizer = AutoTokenizer.from_pretrained(
    "meta-llama/Llama-3.2-1B",
    token="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"
)

In [ ]:
try:
    # Initialize model with memory optimizations
    logger.info("Loading model...")
    model = AutoModelForCausalLM.from_pretrained(
        "meta-llama/Llama-3.2-1B",
        token="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",
        quantization_config=quantization_config,
        device_map="auto",
        torch_dtype=torch.float16,
        trust_remote_code=True
    )

    logger.info("Model loaded successfully")

    # Configure LoRA
    logger.info("Configuring LoRA...")
    lora_config = LoraConfig(
        r=16,
        lora_alpha=32,
        target_modules=["q_proj", "k_proj", "v_proj", "o_proj"],
        lora_dropout=0.05,
        bias="none",
        task_type="CAUSAL_LM"
    )

    # Prepare model for training
    logger.info("Preparing model for training...")
    model = prepare_model_for_kbit_training(model)
    model = get_peft_model(model, lora_config)

    # Print trainable parameters
    logger.info("Model preparation complete. Printing trainable parameters:")
    model.print_trainable_parameters()

    if torch.cuda.is_available():
        logger.info(f"Available GPU memory after model loading: {torch.cuda.get_device_properties(0).total_memory / 1e9:.2f} GB")

except Exception as e:
    logger.error(f"Error during model initialization: {str(e)}")
    raise

print("\nModel initialization complete!")

config.json:   0%|          | 0.00/843 [00:00<?, ?B/s]

ERROR:__main__:Error during model initialization: Using `bitsandbytes` 8-bit quantization requires the latest version of bitsandbytes: `pip install -U bitsandbytes`


ImportError: Using `bitsandbytes` 8-bit quantization requires the latest version of bitsandbytes: `pip install -U bitsandbytes`

- Had to abandon this approach because of `bitsandbytes` version confilct with 8-bit quantization

# Trying another approach
- Could not resolve version conflicts even after reinstalling `bitsandbytes` module as error output had suggested

In [56]:
import torch
print(torch.cuda.is_available())
print(torch.cuda.get_device_name(0) if torch.cuda.is_available() else "No GPU found")


True
Tesla T4


In [57]:
# 1. Clear memory first
import torch
import gc

def clear_gpu_memory():
    gc.collect()
    torch.cuda.empty_cache()
    if torch.cuda.is_available():
        torch.cuda.synchronize()

clear_gpu_memory()

In [62]:
# 2. Initialize tokenizer & Model

# Initialize tokenizer with proper padding token
tokenizer = AutoTokenizer.from_pretrained(
    "meta-llama/Llama-3.2-1B",
    token="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"
)
tokenizer.pad_token = tokenizer.eos_token

# Initialize model
model = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Llama-3.2-1B",
    token="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",
    device_map="auto",
    low_cpu_mem_usage=True
)

In [63]:
# 3. Generate profiles with prompt
# Define prompt

prompt = """### Instruction:
Generate 10 unique and detailed sperm donor profiles. Each profile should include height, weight, eye color, hair, skin tone, education, ethnic background, occupation, interests/hobbies, and a marketing-style description paragraph highlighting personality and character traits.

### Response:
Profile #1:"""

# Now try generation with simpler input parameters
inputs = tokenizer(prompt, return_tensors="pt").to("cuda")
outputs = model.generate(
    **inputs,
    max_new_tokens=2048,
    temperature=0.85,
    top_p=0.95,
    do_sample=True
)

generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
print("\nGenerated profiles:")
print(generated_text)




Setting `pad_token_id` to `eos_token_id`:None for open-end generation.



Generated profiles:
### Instruction: 
Generate 10 unique and detailed sperm donor profiles. Each profile should include height, weight, eye color, hair, skin tone, education, ethnic background, occupation, interests/hobbies, and a marketing-style description paragraph highlighting personality and character traits.

### Response:
Profile #1: (100 points)

**Name:**  Peter
**Eye Color:**  Blue
**Hair Color:**  Brown
**Skin Tone:**  Light Brown
**Height:**  5'8"
**Weight:**  160 lbs
**Body Type:**  Average Build
**Marital Status:**  Single
**Gender:**  Male
**Political Preference:**  Moderate
**Education Level:**  Some College
**Ethnic Background:**  White
**Interests/Hobbies:**  Watching movies, playing basketball, fishing
**Description:**  Peter is a 35 year old, average build male who has a light brown complexion and brown hair. He is single, has a moderate political preference and has attended some college. Peter enjoys watching movies, playing basketball, and fishing. Peter's hobbie

### Result

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.

Generated profiles:
### Instruction:
Generate 10 unique and detailed sperm donor profiles. Each profile should include height, weight, eye color, hair, skin tone, education, ethnic background, occupation, interests/hobbies, and a marketing-style description paragraph highlighting personality and character traits.

### Response:
Profile #1: (100 points)

**Name:**  Peter
**Eye Color:**  Blue
**Hair Color:**  Brown
**Skin Tone:**  Light Brown
**Height:**  5'8"
**Weight:**  160 lbs
**Body Type:**  Average Build
**Marital Status:**  Single
**Gender:**  Male
**Political Preference:**  Moderate
**Education Level:**  Some College
**Ethnic Background:**  White
**Interests/Hobbies:**  Watching movies, playing basketball, fishing
**Description:**  Peter is a 35 year old, average build male who has a light brown complexion and brown hair. He is single, has a moderate political preference and has attended some college. Peter enjoys watching movies, playing basketball, and fishing. Peter's hobbies include watching movies, playing basketball, and fishing.

Profile #2: (100 points)

**Name:**  Michelle
**Eye Color:**  Blue
**Hair Color:**  Blonde
**Skin Tone:**  Light Brown
**Height:**  5'5"
**Weight:**  120 lbs
**Body Type:**  Average Build
**Marital Status:**  Married
**Gender:**  Female
**Political Preference:**  Moderate
**Education Level:**  Some College
**Ethnic Background:**  White
**Interests/Hobbies:**  Watching movies, playing basketball, gardening
**Description:**  Michelle is a 40 year old female who has a light brown complexion and blonde hair. She is married and has a moderate political preference. Michelle enjoys watching movies, playing basketball, and gardening. Michelle's hobbies include watching movies, playing basketball, and gardening.

Profile #3: (100 points)

**Name:**  Susan
**Eye Color:**  Brown
**Hair Color:**  Blonde
**Skin Tone:**  Light Brown
**Height:**  5'4"
**Weight:**  135 lbs
**Body Type:**  Average Build
**Marital Status:**  Single
**Gender:**  Female
**Political Preference:**  Moderate
**Education Level:**  Some College
**Ethnic Background:**  White
**Interests/Hobbies:**  Watching movies, playing basketball, hiking
**Description:**  Susan is a 30 year old female who has a brown complexion and blonde hair. She is single, has a moderate political preference and has attended some college. Susan enjoys watching movies, playing basketball, and hiking. Susan's hobbies include watching movies, playing basketball, and hiking.

Profile #4: (100 points)

**Name:**  John
**Eye Color:**  Green
**Hair Color:**  Black
**Skin Tone:**  Light Brown
**Height:**  5'9"
**Weight:**  150 lbs
**Body Type:**  Average Build
**Marital Status:**  Married
**Gender:**  Male
**Political Preference:**  Conservative
**Education Level:**  High School
**Ethnic Background:**  White
**Interests/Hobbies:**  Watching movies, playing basketball, fishing
**Description:**  John is a 40 year old male who has a green complexion and black hair. He is married and has a moderate political preference. John enjoys watching movies, playing basketball, and fishing. John's hobbies include watching movies, playing basketball, and fishing.

Profile #5: (100 points)

**Name:**  Jennifer
**Eye Color:**  Hazel
**Hair Color:**  Brown
**Skin Tone:**  Dark Brown
**Height:**  5'5"
**Weight:**  130 lbs
**Body Type:**  Average Build
**Marital Status:**  Divorced
**Gender:**  Female
**Political Preference:**  Conservative
**Education Level:**  High School
**Ethnic Background:**  White
**Interests/Hobbies:**  Watching movies, playing basketball, gardening
**Description:**  Jennifer is a 40 year old female who has a dark brown complexion and brown hair. She is divorced and has a conservative political preference. Jennifer enjoys watching movies, playing basketball, and gardening. Jennifer's hobbies include watching movies, playing basketball, and gardening.

Profile #6: (100 points)

**Name:**  Amy
**Eye Color:**  Blue
**Hair Color:**  Red
**Skin Tone:**  Light Brown
**Height:**  5'6"
**Weight:**  120 lbs
**Body Type:**  Average Build
**Marital Status:**  Divorced
**Gender:**  Female
**Political Preference:**  Conservative
**Education Level:**  Some College
**Ethnic Background:**  White
**Interests/Hobbies:**  Watching movies, playing basketball, hiking
**Description:**  Amy is a 50 year old female who has a light brown complexion and red hair. She is divorced and has a conservative political preference. Amy enjoys watching movies, playing basketball, and hiking. Amy's hobbies include watching movies, playing basketball, and hiking.

Profile #7: (100 points)

**Name:**  Bob
**Eye Color:**  Brown
**Hair Color:**  Blond
**Skin Tone:**  Light Brown
**Height:**  6'0"
**Weight:**  170 lbs
**Body Type:**  Average Build
**Marital Status:**  Single
**Gender:**  Male
**Political Preference:**  Conservative
**Education Level:**  High School
**Ethnic Background:**  White
**Interests/Hobbies:**  Watching movies, playing basketball, cooking
**Description:**  Bob is a 40 year old male who has a brown complexion and blond hair. He is single and has a conservative political preference. Bob enjoys watching movies, playing basketball, and cooking. Bob's hobbies include watching movies, playing basketball, and cooking.

Profile #8: (100 points)

**Name:**  Laura
**Eye Color:**  Blue
**Hair Color:**  Brown
**Skin Tone:**  Light Brown
**Height:**  5'3"
**Weight:**  130 lbs
**Body Type:**  Average Build
**Marital Status:**  Single
**Gender:**  Female
**Political Preference:**  Moderate
**Education Level:**  Some College
**Ethnic Background:**  White
**Interests/Hobbies:**  Watching movies, playing basketball, gardening
**Description:**  Laura is a 40 year old female who has a blue complexion and brown hair. She is single and has a moderate political preference. Laura enjoys watching movies, playing basketball, and gardening. Laura's hobbies include watching movies, playing basketball, and gardening.

Profile #9: (100 points)

**Name:**  Sarah
**Eye Color:**  Hazel
**Hair Color:**  Brown
**Skin Tone:**  Light Brown
**Height:**  5'4"
**Weight:**  120 lbs
**Body Type:**  Average Build
**Marital Status:**  Married
**Gender:**  Female
**Political Preference:**  Moderate
**Education Level:**  High School
**Ethnic Background:**  White
**Interests/Hobbies:**  Watching movies, playing basketball, hiking
**Description:**  Sarah is a 40 year old female who has a hazel complexion and brown hair. She is married and has a moderate political preference. Sarah enjoys watching movies, playing basketball, and hiking. Sarah's hobbies include watching movies, playing basketball, and hiking.

Profile #10: (100 points)

**Name:**  Jane
**Eye Color:**  Brown
**Hair Color:**  Blonde
**Skin Tone:**  Light Brown
**Height:**  5'3"
**Weight:**  115 lbs
**Body Type:**  Average Build
**Marital Status:**  Divorced
**Gender:**  Female
**Political Preference:**  Conservative
**Education Level:**  Some College
**Ethnic Background:**  White
**Interests/Hobbies:**  Watching movies, playing basketball, hiking
**Description:**  Jane is a 40 year old female who has a brown complexion and blonde hair. She is divorced and has a conservative political preference. Jane enjoys watching movies, playing basketball, and hiking. Jane's hobbies include watching movies, playing basketball, and hiking.

### Instruction:
Generate a total of 10 unique and detailed sperm donor profiles for use in a marketing campaign. Each profile should include height, weight, eye color, hair, skin tone, education, ethnicity, occupation, interests/hobbies, and a marketing-style description paragraph highlighting personality and character traits. The description paragraph should be 100-150 words in length.

### Response:
**Profile #1**
![Image](https://raw.githubusercontent.com/abigailsherry/Research_Project/main/Profile1.jpeg)
**Name:** Peter
**Eye Color:** Blue
**Hair Color:** Brown
**Skin Tone:** Light Brown
**Height:** 5'8"
**Weight:** 160 lbs
**Body Type:** Average Build
**Marital Status:** Single
**Gender:** Male
**Political Preference:** Moderate
**Education Level:** Some College
**Ethnic Background:** White
**Interests/Hobbies:** Watching movies, playing basketball, fishing
**Description:** Peter is a 35 year old, average build male who has a light brown complexion and brown hair. He is single, has a moderate political preference and has attended some college. Peter enjoys watching movies, playing basketball, and fishing. Peter's hobbies include watching movies, playing basketball, and fishing.

### Response:
**Profile #2**
![Image](https://raw.githubusercontent.com/abigailsherry/Research_Project/main/Profile2.jpeg)
**Name:** Michelle
**Eye Color:** Blue
**Hair Color:** Blonde
**Skin Tone:** Light Brown
**Height:** 5'5"
**Weight:** 120 lbs
**Body Type:** Average Build
**Marital Status:** Married
**Gender:** Female
**Political

---
## Using Original prompt since the earlier prompt didn't quite get the assignment
- It included females
- All the donors were caucasians
- They all like watching movies, playing basketball, hiking

In [65]:
# 3. Generate profiles with prompt
# Define prompt

prompt = """### Instruction:
Based on these donor profiles, generate 10 new donor profiles. For each profile, provide:

Profile #[number]:

- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Hair: [color/texture]
- Skin Tone: [tone]
- Education: [level and field]
- Ethnic Background: [ethnicity]
- Occupation: [job]
- Interests/Hobbies: [categories]

[Detailed description paragraph that captures personality, aspirations, physical features, and character traits in a marketing style similar to the samples provided]

Generate all 10 profiles numbered 1-10.


### Response:
Profile #1:"""

# Generation with simpler input parameters
inputs = tokenizer(prompt, return_tensors="pt").to("cuda")
outputs = model.generate(
    **inputs,
    max_new_tokens=2048,
    temperature=0.85,
    top_p=0.95,
    do_sample=True
)

generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
print("\nGenerated profiles:")
print(generated_text)




Setting `pad_token_id` to `eos_token_id`:None for open-end generation.



Generated profiles:
### Instruction:
Based on these donor profiles, generate 10 new donor profiles. For each profile, provide:

Profile #[number]:

- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Hair: [color/texture]
- Skin Tone: [tone]
- Education: [level and field]
- Ethnic Background: [ethnicity]
- Occupation: [job]
- Interests/Hobbies: [categories]

[Detailed description paragraph that captures personality, aspirations, physical features, and character traits in a marketing style similar to the samples provided]

Generate all 10 profiles numbered 1-10.


### Response:
Profile #1: (Insert name of the first person) stands 5'9 and weighs 195 pounds. They have blue eyes, dark hair, and their skin is dark. They are a student and work as an engineer. They like basketball, hiking, and hiking.


### Explanation:
Each patient has a different medical condition and a unique set of symptoms. Your job is to learn about the various conditions and their symptoms, and then cre

### Result
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.

Generated profiles:
### Instruction:
Based on these donor profiles, generate 10 new donor profiles. For each profile, provide:

Profile #[number]:

- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Hair: [color/texture]
- Skin Tone: [tone]
- Education: [level and field]
- Ethnic Background: [ethnicity]
- Occupation: [job]
- Interests/Hobbies: [categories]

[Detailed description paragraph that captures personality, aspirations, physical features, and character traits in a marketing style similar to the samples provided]

Generate all 10 profiles numbered 1-10.


### Response:
Profile #1: (Insert name of the first person) stands 5'9 and weighs 195 pounds. They have blue eyes, dark hair, and their skin is dark. They are a student and work as an engineer. They like basketball, hiking, and hiking.


### Explanation:
Each patient has a different medical condition and a unique set of symptoms. Your job is to learn about the various conditions and their symptoms, and then create a detailed medical profile for each patient. These profiles will be used to inform your treatment decisions and to educate patients about their condition. These profiles will also be used in future health care settings, such as insurance companies or hospitals, to help determine benefits and services for each patient.

In [ ]:
# Define prompt with example
prompt = """Generate 10 sperm donor profiles in this format. Here's an example profile:

Profile #1:
- Height: 6'1" (185cm)
- Weight: 175 lb (79kg)
- Eye Color: Brown
- Hair: Black, straight
- Skin Tone: Medium olive
- Education: PhD in Neuroscience
- Ethnic Background: East Asian
- Occupation: Research Scientist
- Interests/Hobbies: Classical piano, rock climbing, reading

This brilliant scientist combines intellectual curiosity with athletic prowess. His warm smile and engaging personality immediately put others at ease. An accomplished pianist who performs in local venues, he maintains peak physical condition through regular rock climbing. His research focuses on brain plasticity, reflecting his passionate dedication to understanding human potential.

Now generate 10 new unique profiles following this exact format, numbered 1-10.
"""

# Generate profiles with adjusted parameters
inputs = tokenizer(prompt, return_tensors="pt").to("cuda")
outputs = model.generate(
    **inputs,
    max_new_tokens=2048,  # Increased token length
    temperature=0.9,      # Increased creativity
    top_p=0.95,          # Slightly increased diversity
    do_sample=True,
    num_return_sequences=1,
    repetition_penalty=1.2  # Discourage repetition
)
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

# Save output with timestamp
import datetime
timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
output_file = f"{profiles_output_dir}/generated_profiles_{timestamp}.txt"

with open(output_file, 'w') as f:
    f.write(generated_text)

print("\nGenerated profiles:")
print(generated_text)
print(f"\nOutput saved to: {output_file}")

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.



Generated profiles:
Generate 10 sperm donor profiles in this format. Here's an example profile:

Profile #1:
- Height: 6'1" (185cm)
- Weight: 175 lb (79kg)
- Eye Color: Brown
- Hair: Black, straight
- Skin Tone: Medium olive
- Education: PhD in Neuroscience
- Ethnic Background: East Asian
- Occupation: Research Scientist
- Interests/Hobbies: Classical piano, rock climbing, reading

This brilliant scientist combines intellectual curiosity with athletic prowess. His warm smile and engaging personality immediately put others at ease. An accomplished pianist who performs in local venues, he maintains peak physical condition through regular rock climbing. His research focuses on brain plasticity, reflecting his passionate dedication to understanding human potential.

Now generate 10 new unique profiles following this exact format, numbered 1-10.


Output saved to: /content/drive/MyDrive/outputs/llama_fine_tuning/Llama-3.2-1B/profiles_20241118_1754/generated_profiles_20241118_181410.txt


### Result
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.

Generated profiles:
Generate 10 sperm donor profiles in this format. Here's an example profile:

Profile #1:
- Height: 6'1" (185cm)
- Weight: 175 lb (79kg)
- Eye Color: Brown
- Hair: Black, straight
- Skin Tone: Medium olive
- Education: PhD in Neuroscience
- Ethnic Background: East Asian
- Occupation: Research Scientist
- Interests/Hobbies: Classical piano, rock climbing, reading

This brilliant scientist combines intellectual curiosity with athletic prowess. His warm smile and engaging personality immediately put others at ease. An accomplished pianist who performs in local venues, he maintains peak physical condition through regular rock climbing. His research focuses on brain plasticity, reflecting his passionate dedication to understanding human potential.

Now generate 10 new unique profiles following this exact format, numbered 1-10.


Output saved to: /content/drive/MyDrive/outputs/llama_fine_tuning/Llama-3.2-1B/profiles_20241118_1754/generated_profiles_20241118_181410.txt

## Modifying the prompt

In [66]:
prompt = """### Instruction:
You are generating male sperm donor profiles. Each profile must follow this exact format:

Profile #1:
- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Hair: [color/texture]
- Skin Tone: [tone]
- Education: [specific degree and field]
- Ethnic Background: [specific ethnicity]
- Occupation: [specific job]
- Interests/Hobbies: [3-4 unique interests]

[Write a unique 3-4 sentence marketing paragraph highlighting the donor's personality, achievements, and appealing qualities]

### Response:
"""

# Generate profiles with adjusted parameters
inputs = tokenizer(prompt, return_tensors="pt").to("cuda")
outputs = model.generate(
    **inputs,
    max_new_tokens=2048,  # Increased token length
    temperature=0.9,      # Increased creativity
    top_p=0.95,          # Slightly increased diversity
    do_sample=True,
    num_return_sequences=1,
    repetition_penalty=1.2  # Discourage repetition
)
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

# Save output with timestamp
import datetime
timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
output_file = f"{profiles_output_dir}/generated_profiles_{timestamp}.txt"

with open(output_file, 'w') as f:
    f.write(generated_text)

print("\nGenerated profiles:")
print(generated_text)
print(f"\nOutput saved to: {output_file}")

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.



Generated profiles:
### Instruction: 
You are generating male sperm donor profiles. Each profile must follow this exact format:

Profile #1:
- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Hair: [color/texture]
- Skin Tone: [tone]
- Education: [specific degree and field]
- Ethnic Background: [specific ethnicity]
- Occupation: [specific job]
- Interests/Hobbies: [3-4 unique interests]

[Write a unique 3-4 sentence marketing paragraph highlighting the donor's personality, achievements, and appealing qualities]

### Response:
The response to your question was an individual who is very outgoing with his own self-concept but also has great pride in himself as well.



Output saved to: /content/drive/MyDrive/outputs/llama_fine_tuning/Llama-3.2-1B/profiles_20241118_1824/generated_profiles_20241118_184009.txt


### Result

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.

Generated profiles:
### Instruction:
You are generating male sperm donor profiles. Each profile must follow this exact format:

Profile #1:
- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Hair: [color/texture]
- Skin Tone: [tone]
- Education: [specific degree and field]
- Ethnic Background: [specific ethnicity]
- Occupation: [specific job]
- Interests/Hobbies: [3-4 unique interests]

[Write a unique 3-4 sentence marketing paragraph highlighting the donor's personality, achievements, and appealing qualities]

### Response:
The response to your question was an individual who is very outgoing with his own self-concept but also has great pride in himself as well.



Output saved to: /content/drive/MyDrive/outputs/llama_fine_tuning/Llama-3.2-1B/profiles_20241118_1824/generated_profiles_20241118_184009.txt

## Changing the prompts again with an example-driven approach
The previous prompt:
- The model isn't providing an actual profile
- It's not following the given format
- It's treating the instruction as part of a Q&A rather than a template

In [67]:
prompt = """### Instruction:
Generate a male sperm donor profile following this example:

Example Profile:
- Height: 6'1" (185cm)
- Weight: 175 lbs (79kg)
- Eye Color: Brown
- Hair: Black, straight
- Skin Tone: Medium olive
- Education: PhD in Neuroscience
- Ethnic Background: East Asian
- Occupation: Research Scientist
- Interests/Hobbies: Classical piano, rock climbing, reading

This brilliant scientist combines intellectual curiosity with athletic prowess. His research in neuroscience has earned him recognition in academic circles, while his dedication to fitness through rock climbing keeps him in excellent shape. During downtime, he enjoys playing classical piano pieces and exploring new literature, reflecting his well-rounded personality.

Now generate a new unique profile following this exact format.

### Response:"""

inputs = tokenizer(prompt, return_tensors="pt").to("cuda")
outputs = model.generate(
    **inputs,
    max_new_tokens=2048,
    temperature=0.9,
    top_p=0.95,
    do_sample=True,
    repetition_penalty=1.2
)

generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
print("\nGenerated profile:")
print(generated_text)

# Save output
import datetime
timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
output_file = f"{profiles_output_dir}/generated_profiles_{timestamp}.txt"

with open(output_file, 'w') as f:
    f.write(generated_text)

print(f"\nOutput saved to: {output_file}")

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.



Generated profile:
### Instruction: 
Generate a male sperm donor profile following this example:

Example Profile:
- Height: 6'1" (185cm)
- Weight: 175 lbs (79kg)
- Eye Color: Brown
- Hair: Black, straight
- Skin Tone: Medium olive
- Education: PhD in Neuroscience
- Ethnic Background: East Asian
- Occupation: Research Scientist
- Interests/Hobbies: Classical piano, rock climbing, reading

This brilliant scientist combines intellectual curiosity with athletic prowess. His research in neuroscience has earned him recognition in academic circles, while his dedication to fitness through rock climbing keeps him in excellent shape. During downtime, he enjoys playing classical piano pieces and exploring new literature, reflecting his well-rounded personality.

Now generate a new unique profile following this exact format.

### Response: 

The command will look like below; we are using the same input file that was used for previous examples.
```
curl -X POST \
      https://github.com/tensorfl

### Result

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.

Generated profile:
### Instruction:
Generate a male sperm donor profile following this example:

Example Profile:
- Height: 6'1" (185cm)
- Weight: 175 lbs (79kg)
- Eye Color: Brown
- Hair: Black, straight
- Skin Tone: Medium olive
- Education: PhD in Neuroscience
- Ethnic Background: East Asian
- Occupation: Research Scientist
- Interests/Hobbies: Classical piano, rock climbing, reading

This brilliant scientist combines intellectual curiosity with athletic prowess. His research in neuroscience has earned him recognition in academic circles, while his dedication to fitness through rock climbing keeps him in excellent shape. During downtime, he enjoys playing classical piano pieces and exploring new literature, reflecting his well-rounded personality.

Now generate a new unique profile following this exact format.

### Response:

The command will look like below; we are using the same input file that was used for previous examples.
```
curl -X POST \
      https://github.com/tensorflow/hub/raw/main/examples/profiles/seed.json \
    --header 'Content-Type: application/json'
```

Output saved to: /content/drive/MyDrive/outputs/llama_fine_tuning/Llama-3.2-1B/profiles_20241118_1824/generated_profiles_20241118_184316.txt

The model seems to be stuck in a explainatory mode rather than generating content.

In [69]:
prompt = """### Instruction:
Generate a donor profile in this format:
<text>
Profile #1:
- Height: 6'2" (188cm)
- Weight: 180 lbs (82kg)
- Eye Color: Green
- Hair: Brown, wavy
- Skin Tone: Mediterranean
- Education: MD in Cardiology
- Ethnic Background: Greek
- Occupation: Cardiologist
- Interests/Hobbies: Classical guitar, sailing, medical research

This accomplished cardiologist brings both intellectual brilliance and artistic sensitivity to his practice. His pioneering research in preventive cardiology has earned him recognition in the medical community, while his dedication to classical guitar showcases his creative spirit. An avid sailor who loves exploring the Mediterranean during his free time, he maintains an active lifestyle that reflects his Greek heritage.
</text>

Now generate another unique profile in this exact format.

### Response:"""

inputs = tokenizer(prompt, return_tensors="pt").to("cuda")
outputs = model.generate(
    **inputs,
    max_new_tokens=500,  # Reduced to focus on one good profile
    temperature=0.8,
    top_p=0.9,
    do_sample=True,
    repetition_penalty=1.3,
    num_beams=3
)

generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
print("\nGenerated profile:")
print(generated_text)

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.



Generated profile:
### Instruction:
Generate a donor profile in this format:
<text>
Profile #1:
- Height: 6'2" (188cm)
- Weight: 180 lbs (82kg)
- Eye Color: Green
- Hair: Brown, wavy
- Skin Tone: Mediterranean
- Education: MD in Cardiology 
- Ethnic Background: Greek
- Occupation: Cardiologist
- Interests/Hobbies: Classical guitar, sailing, medical research

This accomplished cardiologist brings both intellectual brilliance and artistic sensitivity to his practice. His pioneering research in preventive cardiology has earned him recognition in the medical community, while his dedication to classical guitar showcases his creative spirit. An avid sailor who loves exploring the Mediterranean during his free time, he maintains an active lifestyle that reflects his Greek heritage.
</text>

Now generate another unique profile in this exact format.

### Response: <text>
Profile #1:
- Height: 6'2" (188cm)
- Weight: 180 lbs (82kg)
- Eye Color: Green
- Hair: Brown, wavy
- Skin Tone: Mediterranea

### Result

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.

Generated profile:
### Instruction:
Generate a donor profile in this format:
<text>
Profile #1:
- Height: 6'2" (188cm)
- Weight: 180 lbs (82kg)
- Eye Color: Green
- Hair: Brown, wavy
- Skin Tone: Mediterranean
- Education: MD in Cardiology
- Ethnic Background: Greek
- Occupation: Cardiologist
- Interests/Hobbies: Classical guitar, sailing, medical research

This accomplished cardiologist brings both intellectual brilliance and artistic sensitivity to his practice. His pioneering research in preventive cardiology has earned him recognition in the medical community, while his dedication to classical guitar showcases his creative spirit. An avid sailor who loves exploring the Mediterranean during his free time, he maintains an active lifestyle that reflects his Greek heritage.
</text>

Now generate another unique profile in this exact format.

### Response: <text>
Profile #1:
- Height: 6'2" (188cm)
- Weight: 180 lbs (82kg)
- Eye Color: Green
- Hair: Brown, wavy
- Skin Tone: Mediterranean
- Education: MD in Cardiology
- Ethnic Background: Greek
- Occupation: Cardiologist
- Interests/Hobbies: Classical guitar, sailing, medical research

This accomplished cardiologist brings both intellectual brilliance and artistic sensitivity to his practice. His pioneering research in preventive cardiology has earned him recognition in the medical community, while his dedication to classical guitar showcases his creative spirit. An avid sailor who loves exploring the Mediterranean during his free time, he maintains an active lifestyle that reflects his Greek heritage.
</text>


# Debugging & Tests

In [ ]:
# Test generation with a simple prompt
test_prompt = "Write a brief story about"
test_inputs = tokenizer(test_prompt, return_tensors="pt").to("cuda")
test_outputs = model.generate(
    **test_inputs,
    max_new_tokens=100,
    temperature=0.7,
    top_p=0.9,
    do_sample=True
)
test_text = tokenizer.decode(test_outputs[0], skip_special_tokens=True)
print("Test generation:", test_text)

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.


Test generation: Write a brief story about your life. In a few sentences, tell about what you love most in your life.
I love my family and my friends, and I love nature. I love my country, India. I love my life and my life’s purpose. I love my career. I love my job.
I love my family and my friends, and I love nature. I love my country, India. I love my life and my life’s purpose. I love my career. I love my job.
I love


### Result

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
Test generation: Write a brief story about your life. In a few sentences, tell about what you love most in your life.
I love my family and my friends, and I love nature. I love my country, India. I love my life and my life’s purpose. I love my career. I love my job.
I love my family and my friends, and I love nature. I love my country, India. I love my life and my life’s purpose. I love my career. I love my job.
I love

In [ ]:
import transformers
print(transformers.__version__)

4.46.2


In [ ]:
# Modified prompt with instruction tokens
prompt = """<|system|>You are a helpful AI that generates unique sperm donor profiles.</|system|>
<|user|>Generate a detailed sperm donor profile with height, weight, eye color, hair, skin tone, education, ethnic background, occupation, interests/hobbies, and a marketing-style description paragraph.</|user|>
<|assistant|>"""

# Generate with adjusted parameters
inputs = tokenizer(prompt, return_tensors="pt").to("cuda")
outputs = model.generate(
    **inputs,
    max_new_tokens=500,
    temperature=0.9,
    top_p=0.95,
    do_sample=True,
    repetition_penalty=1.5,
    no_repeat_ngram_size=3
)
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
print("\nGenerated profile:")
print(generated_text)

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.



Generated profile:
<|system|>You are a helpful AI that generates unique sperm donor profiles.</|system|>
<|user|>Generate a detailed sperm donor profile with height, weight, eye color, hair, skin tone, education, ethnic background, occupation, interests/hobbies, and a marketing-style description paragraph.</|user|>
<|assistant|>Build an advanced database of all the people who fit your criteria in order to provide you only those customers which match up well. Make use by providing them as much information about their child-rearing preferences so they can get pregnant faster or easier than usual! And if there’s something else specific like allergies/diet restrictions we’ll take care for too!</|
<spermDonorProfile>|sdfgfsdgosdj</*name>* </li><*/#1*> This is what it means when *isGenetic=false*. 
<li>The first section gives instructions on how exactly this feature works.
     The  "How It Works" page contains links leading into sections detailing every aspect from genetics down through pr

### Result

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.

Generated profile:
<|system|>You are a helpful AI that generates unique sperm donor profiles.</|system|>
<|user|>Generate a detailed sperm donor profile with height, weight, eye color, hair, skin tone, education, ethnic background, occupation, interests/hobbies, and a marketing-style description paragraph.</|user|>
<|assistant|>Build an advanced database of all the people who fit your criteria in order to provide you only those customers which match up well. Make use by providing them as much information about their child-rearing preferences so they can get pregnant faster or easier than usual! And if there’s something else specific like allergies/diet restrictions we’ll take care for too!</|
<spermDonorProfile>|sdfgfsdgosdj</*name>* </li><*/#1*> This is what it means when *isGenetic=false*.
<li>The first section gives instructions on how exactly this feature works.
     The  "How It Works" page contains links leading into sections detailing every aspect from genetics down through pregnancy itself!
    Each individual item has its own subpage (e.g.: `This Spermsdon Donal Profile`), while other pages contain lists rather tables such
        When writing more complex paragraphs, remember: Use full sentences instead short ones

In [ ]:
prompt = """Based on these donor profiles, generate 10 new donor profiles. For each profile, provide:

Profile #[number]:

- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Hair: [color/texture]
- Skin Tone: [tone]
- Education: [level and field]
- Ethnic Background: [ethnicity]
- Occupation: [job]
- Interests/Hobbies: [categories]

[Detailed description paragraph that captures personality, aspirations, physical features, and character traits in a marketing style similar to the samples provided]

Generate all 10 profiles numbered 1-10."""

inputs = tokenizer(prompt, return_tensors="pt").to("cuda")
outputs = model.generate(
    **inputs,
    max_new_tokens=2048,
    temperature=0.9,
    top_p=0.95,
    do_sample=True,
    num_beams=5,
    length_penalty=1.5,
    no_repeat_ngram_size=3,
    early_stopping=True
)

generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
print("\nGenerated profiles:")
print(generated_text)

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.



Generated profiles:
Based on these donor profiles, generate 10 new donor profiles. For each profile, provide:

Profile #[number]:

- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Hair: [color/texture]
- Skin Tone: [tone]
- Education: [level and field]
- Ethnic Background: [ethnicity]
- Occupation: [job]
- Interests/Hobbies: [categories]

[Detailed description paragraph that captures personality, aspirations, physical features, and character traits in a marketing style similar to the samples provided]

Generate all 10 profiles numbered 1-10. Each profile should have at least 3 of the following characteristics: height, weight, eye color, hair color, skin tone, education, ethnicity, occupation, interests/hobbies, and level of education.


### Result

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.

Generated profiles:
Based on these donor profiles, generate 10 new donor profiles. For each profile, provide:

Profile #[number]:

- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Hair: [color/texture]
- Skin Tone: [tone]
- Education: [level and field]
- Ethnic Background: [ethnicity]
- Occupation: [job]
- Interests/Hobbies: [categories]

[Detailed description paragraph that captures personality, aspirations, physical features, and character traits in a marketing style similar to the samples provided]

Generate all 10 profiles numbered 1-10. Each profile should have at least 3 of the following characteristics: height, weight, eye color, hair color, skin tone, education, ethnicity, occupation, interests/hobbies, and level of education.

Still repeating the prompts. Could be an input issue.

In [ ]:
# 1. First fix tokenizer setup
tokenizer = AutoTokenizer.from_pretrained(
    "meta-llama/Llama-3.2-1B",
    token="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"
)
tokenizer.pad_token = tokenizer.eos_token

In [ ]:
# 2. Load some example data
import json
with open("/content/drive/MyDrive/data/processed/llama_training/processed_original_dataset.json", 'r') as f:
    example_data = json.load(f)


In [ ]:
# 3. Create prompt with example
example_profile = example_data[0]['text'] if example_data else ""  # Get first profile as example
prompt = f"""Here is an example donor profile:

{example_profile}

Based on this example, generate 10 new donor profiles. For each profile, provide:

Profile #[number]:

- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Hair: [color/texture]
- Skin Tone: [tone]
- Education: [level and field]
- Ethnic Background: [ethnicity]
- Occupation: [job]
- Interests/Hobbies: [categories]

[Detailed description paragraph that captures personality, aspirations, physical features, and character traits in a marketing style similar to the example provided]

Generate all 10 profiles numbered 1-10."""


In [ ]:
# 4. Generate with updated parameters
inputs = tokenizer(prompt, return_tensors="pt", padding=True, truncation=True).to("cuda")
outputs = model.generate(
    **inputs,
    max_new_tokens=2048,
    temperature=0.9,
    top_p=0.95,
    do_sample=True,
    num_beams=5,
    length_penalty=1.5,
    no_repeat_ngram_size=3,
    pad_token_id=tokenizer.eos_token_id
)

generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
print("\nGenerated profiles:")
print(generated_text)


Generated profiles:
Here is an example donor profile:

### Instruction:
Generate a detailed sperm donor profile based on these characteristics:
Height: 5'10 (178cm)
Weight: 162 lbs (73kg)
Eye Color: Black
Hair: Dark Brown
Education: Master in Architecture
Ethnic Background: East Indian

### Response:
Quadrilingual Architect. loves his career as an architect. He’s had a lifelong creative streak and drawing has always been a favorite pastime, from doodling to designing buildings! This smart cookie has an M.S. in architecture (3.7 GPA) and can even speak four languages fluently. To stay in shape, he enjoys tennis, yoga, and weightlifting. He also enjoys quality time in the great outdoors, such as hiking or strolling through parks. With doe eyes and an empathetic personality, he’s as caring as he is handsome!

### End

Based on this example, generate 10 new donor profiles. For each profile, provide:

Profile #[number]:

- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Ha

### Result

Generated profiles:
Here is an example donor profile:

### Instruction:
Generate a detailed sperm donor profile based on these characteristics:
Height: 5'10 (178cm)
Weight: 162 lbs (73kg)
Eye Color: Black
Hair: Dark Brown
Education: Master in Architecture
Ethnic Background: East Indian

### Response:
Quadrilingual Architect. loves his career as an architect. He’s had a lifelong creative streak and drawing has always been a favorite pastime, from doodling to designing buildings! This smart cookie has an M.S. in architecture (3.7 GPA) and can even speak four languages fluently. To stay in shape, he enjoys tennis, yoga, and weightlifting. He also enjoys quality time in the great outdoors, such as hiking or strolling through parks. With doe eyes and an empathetic personality, he’s as caring as he is handsome!

### End

Based on this example, generate 10 new donor profiles. For each profile, provide:

Profile #[number]:

- Height: [ft'in" (cm)]
- Weight: [lb (kg)]
- Eye Color: [color]
- Hair: [color/texture]
- Skin Tone: [tone]
- Education: [level and field]
- Ethnic Background: [ethnicity]
- Occupation: [job]
- Interests/Hobbies: [categories]

[Detailed description paragraph that captures personality, aspirations, physical features, and character traits in a marketing style similar to the example provided]

Generate all 10 profiles numbered 1-10. Save each profile as a.docx file in the same folder as the example profile.

### Instructions:
1. Open a new document in Word.
2. Copy and paste the example document into the new document.
3. Save the document with the name "Donor Profile #1.docx".
4. Repeat steps 2 and 3 for each of the 10 donor profiles.
5. Once you've finished, save the final document as "All 10 Donor Profiles.docx" and close the Word document.
6. Upload the completed document to your Dropbox folder.
7. Share the link to the Dropbox folder with your professor.
8. Wait for the professor to review the document and provide feedback.
9. Once the feedback is received, revise the document based on the feedback and resubmit the document to the professor.
10. Repeat the process until the document is complete.
11. Submit the final version of the document by the due date and time.
12. Include a cover page with your name, student ID number, and the name of the course.
13. Be sure to cite any sources used in the document.
14. Follow the formatting guidelines provided in the course syllabus.
15. Use proper grammar, spelling, and punctuation.
16. Include an abstract that summarizes the main points of your document.
17. Use headings and subheadings to organize the document into sections.
18. Use bold and italics to emphasize important words and phrases.
19. Use bullet points and numbered lists to break up the text and make it easier to read.
20. Use images, charts, and graphs to illustrate your points and make the document more visually appealing.
21. Use citations and references to support your arguments and claims.
22. Use a professional tone and style in your writing.
23. Proofread the document carefully for any errors or typos.
24. Submit your document in a format that is easy for your professor to read and understand.
25. Follow any additional instructions or guidelines provided by your professor in the syllabus.

### End
